### PR TITLE
Adding new event to trigger handler to store last login time after successful federated authentication.

### PR DESCRIPTION
**Pull request by ZiyamSanthosh**
### Proposed changes in this pull request

- A new event has been introduced and fired at the end of JITProvisioningPostAuthenticationHandler to store the last login time of the associated local user.

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java [hunk 1]
```diff
@@ -230,6 +234,9 @@ private PostAuthnHandlerFlowStatus handleResponseFlow(HttpServletRequest request
                     callDefaultProvisioningHandler(username, context, externalIdPConfig, combinedLocalClaims,
                             stepConfig);
                     handleConsents(request, stepConfig, context.getTenantDomain());
+                    // Trigger event to store last login time after successful authentication.
+                    String emailUsername = localClaimValues.get(EMAIL_ADDRESS_CLAIM);
+                    triggerEvent(username, context, emailUsername);
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java:239]: why is this coupled to email claims specifically? username can be non-email as well.

File: components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java [hunk 2]
```diff
@@ -200,6 +200,7 @@ private Event(){}
         public static final String POST_GENERATE_SMS_OTP = "POST_GENERATE_SMS_OTP";
         public static final String POST_VALIDATE_SMS_OTP = "POST_VALIDATE_SMS_OTP";
         public static final String SESSION_EXTENSION = "SESSION_EXTENSION";
+        public static final String POST_JIT_USER_UPDATE = "POST_JIT_USER_UPDATE";
```
Comment (pulasthi7) [inline components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java:203]: why only the update event is considered? what about the create event?

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java (hunk 1 above)
Comment (ZiyamSanthosh) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java]: Fixed in https://github.com/wso2/carbon-identity-framework/pull/4612/commits/a4e3a9878fea0d2b572975f81dffafd696fbe36c

File: components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java (hunk 2 above)
Comment (ZiyamSanthosh) [reply components/identity-event/org.wso2.carbon.identity.event/src/main/java/org/wso2/carbon/identity/event/IdentityEventConstants.java]: This handler gets triggered even if the user already provisioned or not. So decided to fire the event from here. The event name might confuse. So, changed it from `POST_JIT_USER_UPDATE` to `JIT_USER_SET_CLAIMS`. Fixed in https://github.com/wso2/carbon-identity-framework/pull/4612/commits/a4e3a9878fea0d2b572975f81dffafd696fbe36c

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java [hunk 3]
```diff
@@ -426,12 +433,49 @@ private PostAuthnHandlerFlowStatus handleRequestFlow(HttpServletRequest request,
...
+                    .getProvisioningUserStoreId(), realm, emailUsername);
+            IdentityEventService eventService = FrameworkServiceDataHolder.getInstance().getIdentityEventService();
+            Map<String, Object> eventProperties = new HashMap<>();
+            eventProperties.put(IdentityEventConstants.EventProperty.USER_NAME, username);
+            eventProperties.put(IdentityEventConstants.EventProperty.USER_STORE_MANAGER, userStoreManager);
+            eventProperties.put(IdentityEventConstants.EventProperty.OPERATION_STATUS, true);
+            Event event = new Event(IdentityEventConstants.Event.JIT_USER_SET_CLAIMS, eventProperties);
+            eventService.handleEvent(event);
+        } catch (UserStoreException e) {
+            // TODO: change this to a proper error code.
```
Comment (ZiyamSanthosh) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java:468]: Remove comment

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java [hunk 4]
```diff
@@ -426,12 +433,49 @@ private PostAuthnHandlerFlowStatus handleRequestFlow(HttpServletRequest request,
...
+        try {
+            UserRealm realm = getUserRealm(context.getTenantDomain());
+            UserStoreManager userStoreManager = getUserStoreManager(context.getExternalIdP()
+                    .getProvisioningUserStoreId(), realm, emailUsername);
+            IdentityEventService eventService = FrameworkServiceDataHolder.getInstance().getIdentityEventService();
+            Map<String, Object> eventProperties = new HashMap<>();
+            eventProperties.put(IdentityEventConstants.EventProperty.USER_NAME, username);
+            eventProperties.put(IdentityEventConstants.EventProperty.USER_STORE_MANAGER, userStoreManager);
+            eventProperties.put(IdentityEventConstants.EventProperty.OPERATION_STATUS, true);
+            Event event = new Event(IdentityEventConstants.Event.JIT_USER_SET_CLAIMS, eventProperties);
```
Comment (pulasthi7) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java:465]: change to two events one for Jit user add, and one for Jit user update. and fire those events separately. In the event handler, you may treat the two events in the same manner for this feature

Comment (ZiyamSanthosh) [inline components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java:465]: Change the event name into POST_USER_JIT_PROVISION

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java (hunk 3 above)
Comment (ZiyamSanthosh) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java]: Fixed in [6643361](https://github.com/wso2/carbon-identity-framework/pull/4612/commits/6643361a2944594606c0a5a677c9580fbfd3eeef)

File: components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java (hunk 4 above)
Comment (ZiyamSanthosh) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java]: Decided to go with the current implementation

Comment (ZiyamSanthosh) [reply components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/JITProvisioningPostAuthenticationHandler.java]: Fixed in https://github.com/wso2/carbon-identity-framework/pull/4612/commits/6643361a2944594606c0a5a677c9580fbfd3eeef

**Review by lashinijay — APPROVED** [review]

**Pull request merged**